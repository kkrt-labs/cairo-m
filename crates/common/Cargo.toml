[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
