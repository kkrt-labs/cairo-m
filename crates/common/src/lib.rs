pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
