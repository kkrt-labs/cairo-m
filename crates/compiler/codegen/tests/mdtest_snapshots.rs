//! Codegen snapshot tests for mdtest cases.
//! This file automatically generates codegen snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the entire compilation pipeline.

mod common;

use cairo_m_compiler_codegen::CodeGenerator;
use cairo_m_compiler_mir::generate_mir;
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_codegen_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("CASM", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            // First generate MIR
            let mir_module = match generate_mir(&db, crate_id) {
                Ok(module) => module,
                Err(diagnostics) => {
                    return Err(format!("MIR generation failed: {:#?}", diagnostics));
                }
            };

            // Then generate CASM code from MIR
            let mut generator = CodeGenerator::new();
            match generator.generate_module(&mir_module) {
                Ok(_) => Ok(generator.debug_instructions()),
                Err(e) => Err(format!("Code generation failed: {:#?}", e)),
            }
        })
        .with_parent_dir(true);

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("Codegen snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
