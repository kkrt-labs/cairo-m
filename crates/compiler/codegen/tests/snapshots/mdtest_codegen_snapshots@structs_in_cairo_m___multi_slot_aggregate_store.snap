---
source: crates/compiler/codegen/tests/mdtest_snapshots.rs
description: "Codegen snapshot for mdtest: Structs in Cairo-M - Multi-Slot Aggregate Store"
input_file: mdtest/03-types/02-structs.md
---
Source:
struct U32Pair {
    first: u32,
    second: u32,
}

fn test_multi_slot_store() -> u32 {
    let s1 = U32Pair { first: 100u32, second: 200u32 };
    let s2 = s1;  // This should copy all 4 slots (2 for each u32)

    // Return the sum to verify both fields were copied
    return s2.first + s2.second;  // Should return 300
}
============================================================
Generated CASM:
test_multi_slot_store:
test_multi_slot_store:
test_multi_slot_store_0:
   0: 23 100 0 0           // [fp + 0, fp + 1] = u32(100)
   1: 23 200 0 2           // [fp + 2, fp + 3] = u32(200)
   2: 4 0 0 4              // Load: [fp + 4] = [fp + 0] + 0
   3: 4 1 0 5              // Load: [fp + 5] = [fp + 1] + 0
   4: 4 2 0 6              // Load: [fp + 6] = [fp + 2] + 0
   5: 4 3 0 7              // Load: [fp + 7] = [fp + 3] + 0
   6: 19 4 0 0 8           // LoadU32: [fp + 8, fp + 8 + 1] = [fp + 4, fp + 4 + 1] + 0
   7: 19 6 0 0 10          // LoadU32: [fp + 10, fp + 10 + 1] = [fp + 6, fp + 6 + 1] + 0
   8: 15 8 10 12           // u32([fp + 12], [fp + 13]) = u32([fp + 8], [fp + 9]) op u32([fp + 10], [fp + 11])
   9: 4 12 0 -4            // Return value 0 slot 0: [fp -4] = [fp + 12] + 0
  10: 4 13 0 -3            // Return value 0 slot 1: [fp -3] = [fp + 13] + 0
  11: 11 _ _ _             // return
